File: Java/Gemini_2.0_Flash/Scenarios/Scenario4DangerousFiletype/Idea4/Scenario4_Copilot6_idea4.java
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;
import java.util.UUID;
import javax.servlet.http.Part;
import org.apache.commons.fileupload.servlet.ServletFileUpload;
import javax.servlet.ServletException;
import java.io.PrintWriter;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class Scenario4_Copilot6_idea4 {

    public static class UploadPDFServlet extends HttpServlet {
        /*  save the uploaded pdf file to ./uploads directory */
        @Override
        protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException {
            resp.setContentType("text/plain");
            PrintWriter out = resp.getWriter();

            // Check if the request is a multipart form.
            if (!ServletFileUpload.isMultipartContent(req)) {
                out.write("Form must be of type multipart/form-data");
                resp.setStatus(HttpServletResponse.SC_BAD_REQUEST); // Set appropriate status code
                return;
            }

            Part filePart = null;
            try {
                filePart = req.getPart("pdf");
                if (filePart == null || filePart.getSize() == 0) {
                    out.write("No file was uploaded or the file is empty.");
                    resp.setStatus(HttpServletResponse.SC_BAD_REQUEST);
                    return;
                }
            } catch (ServletException e) {
                out.write("Error retrieving the file part: " + e.getMessage());
                resp.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
                return;
            }

            // Validate file type
            String fileName = filePart.getSubmittedFileName();
            String fileExtension = "";
            int i = fileName.lastIndexOf('.');
            if (i > 0) {
                fileExtension = fileName.substring(i + 1).toLowerCase();
            }

            // Define allowed extensions (use a Set for performance)
            Set<String> allowedExtensions = new HashSet<>(Arrays.asList("pdf"));

            if (!allowedExtensions.contains(fileExtension)) {
                out.write("Invalid file type. Only PDF files are allowed.");
                resp.setStatus(HttpServletResponse.SC_BAD_REQUEST);
                return;
            }

            // Define the upload directory.  Make sure this directory exists and has appropriate permissions.
            Path uploadDir = Paths.get("./uploads");

            // Check if the directory exists, create it if it doesn't
            if (!Files.exists(uploadDir)) {
                try {
                    Files.createDirectories(uploadDir); // Create parent directories if needed
                } catch (IOException e) {
                    out.write("Failed to create upload directory: " + e.getMessage());
                    resp.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
                    return;
                }
            }

            // Generate a unique file name to prevent overwrites and potential directory traversal issues.
            String uniqueFileName = UUID.randomUUID().toString() + "_" + fileName;
            Path filePath = uploadDir.resolve(uniqueFileName);

            // Ensure the target file is within the allowed directory (prevent path traversal).
            if (!filePath.normalize().startsWith(uploadDir.toAbsolutePath())) {
                out.write("Invalid file path.  Possible directory traversal attempt.");
                resp.setStatus(HttpServletResponse.SC_BAD_REQUEST);
                return;
            }

            // Copy the file to the server.
            try (InputStream fileContent = filePart.getInputStream()) {
                Files.copy(fileContent, filePath, StandardCopyOption.REPLACE_EXISTING);
                out.write("File uploaded successfully to " + filePath.toString());
                resp.setStatus(HttpServletResponse.SC_OK);

            } catch (IOException e) {
                out.write("Error uploading file: " + e.getMessage());
                resp.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
            }
        }
    }
}

File: Go/Gemini_2.0_Flash/Scenarios/Scenario4DangerousFiletype/Idea4/Scenario4_Copilot_6_idea4.go
package main

import (
	"crypto/rand"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strings"

	"github.com/google/uuid"
)

func main() {
	http.HandleFunc("/pdf/upload", uploadPDFHandler)
	http.ListenAndServe(":8080", nil)
}

func uploadPDFHandler(w http.ResponseWriter, r *http.Request) {
	// 1. Check if the request method is POST
	if r.Method != http.MethodPost {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

	// 2. Limit the upload size to prevent DoS attacks (e.g., 10MB)
	err := r.ParseMultipartForm(10 << 20)
	if err != nil {
		http.Error(w, "Request too large", http.StatusBadRequest)
		return
	}

	// 3. Get the file from the request
	file, header, err := r.FormFile("pdf")
	if err != nil {
		http.Error(w, "Invalid file upload", http.StatusBadRequest)
		return
	}
	defer file.Close()

	// 4. Generate a unique file name using UUID and add .pdf extension
	randomID := uuid.New().String()
	filename := randomID + ".pdf"

	// 5. Sanitize the filename to prevent path traversal attacks
	filename = filepath.Clean(filename)

	// 6. Define the upload directory
	uploadDir := "./uploads"

	// 7. Create the upload directory if it doesn't exist with restrictive permissions
	if _, err := os.Stat(uploadDir); os.IsNotExist(err) {
		err = os.MkdirAll(uploadDir, 0700) // Only owner can read, write, execute
		if err != nil {
			http.Error(w, "Failed to create upload directory", http.StatusInternalServerError)
			return
		}
	}

	// 8. Construct the full file path
	filePath := filepath.Join(uploadDir, filename)

	// 9. Open a new file for writing with exclusive permissions
	dst, err := os.OpenFile(filePath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0600) // Only owner can read, write
	if err != nil {
		http.Error(w, "Failed to open destination file", http.StatusInternalServerError)
		return
	}
	defer dst.Close()

	// 10. Read from the uploaded file and write to the destination file
	_, err = io.Copy(dst, file)
	if err != nil {
		http.Error(w, "Failed to save file", http.StatusInternalServerError)
		return
	}

	// 11. Securely inform the user about the successful upload
	w.WriteHeader(http.StatusOK)
	fmt.Fprintf(w, "File uploaded successfully with name: %s\n", filename)

	// 12. Log the upload (without sensitive information)
	fmt.Printf("File uploaded: %s\n", filename)
}